x/staking/keeper/liquid_stake.go

```diff
@@ -153,3 +155,110 @@ func (k Keeper) SafelyDecreaseValidatorBond(ctx sdk.Context, validator types.Val
+// The tokenize share lock store is implemented by keying on the account address
+// and storing a timestamp as the value. The timestamp is empty when the lock is
+// set and gets populated with the unlock completion time once the unlock has started
+func (k Keeper) AddTokenizeSharesLock(ctx sdk.Context, address sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetTokenizeSharesLockKey(address)
+	store.Set(key, sdk.FormatTimeBytes(time.Time{}))
+}
+
+// Removes the tokenize share lock for an account to enable tokenizing shares
+func (k Keeper) RemoveTokenizeSharesLock(ctx sdk.Context, address sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetTokenizeSharesLockKey(address)
+	store.Delete(key)
+}
+
+// Updates the timestamp associated with a lock to the time at which the lock expires
+func (k Keeper) SetTokenizeShareUnlockTime(ctx sdk.Context, address sdk.AccAddress, completionTime time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetTokenizeSharesLockKey(address)
+	store.Set(key, sdk.FormatTimeBytes(completionTime))
+}
+
+// Checks if there is currently a tokenize share lock for a given account
+// Returns a bool indicating if the account is locked, as well as the unlock time
+// which may be empty if an unlock has not been initiated
+func (k Keeper) IsTokenizeSharesDisabled(ctx sdk.Context, address sdk.AccAddress) (disabled bool, unlockTime time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetTokenizeSharesLockKey(address)
+	bz := store.Get(key)
+	if len(bz) == 0 {
+		return false, unlockTime
+	}
+	unlockTime, err := sdk.ParseTimeBytes(bz)
+	if err != nil {
+		panic(err)
+	}
+	return true, unlockTime
+}
+
+// Stores a list of addresses pending tokenize share unlocking at the same time
+func (k Keeper) SetPendingTokenizeShareAuthorizations(ctx sdk.Context, completionTime time.Time, authorizations types.PendingTokenizeShareAuthorizations) {
+	store := ctx.KVStore(k.storeKey)
+	timeKey := types.GetTokenizeShareAuthorizationTimeKey(completionTime)
+	bz := k.cdc.MustMarshal(&authorizations)
+	store.Set(timeKey, bz)
+}
+
+// Returns a list of addresses pending tokenize share unlocking at the same time
+func (k Keeper) GetPendingTokenizeShareAuthorizations(ctx sdk.Context, completionTime time.Time) types.PendingTokenizeShareAuthorizations {
+	store := ctx.KVStore(k.storeKey)
+
+	timeKey := types.GetTokenizeShareAuthorizationTimeKey(completionTime)
+	bz := store.Get(timeKey)
+
+	authorizations := types.PendingTokenizeShareAuthorizations{Addresses: []string{}}
+	if len(bz) == 0 {
+		return authorizations
+	}
+	k.cdc.MustUnmarshal(bz, &authorizations)
+
+	return authorizations
+}
+
+// Inserts thea ddress into a queue where it will sit for 1 unbonding period
+// before the tokenize share lock is removed
+// Returns the completion time
+func (k Keeper) QueueTokenizeSharesAuthorization(ctx sdk.Context, address sdk.AccAddress) time.Time {
```

Comment:
Why do we need a queue, rather than just iterating them all and checking the finish time against the current time? For performance?

Reply:
Yeah main reason is performance. That's how unbondings are done, which I think is pretty slick.
Fwiw, the implementation isn't much more complex than just storing a list.

Reply:
nice, makes sense to me, let's keep it